web/ui/react-app/src/api/Fetcher.tsx

```diff
@@ -0,0 +1,68 @@
+import { Component } from 'react';
+
+export interface FetcherState<T> {
+  data?: T;
+  error?: Error;
+}
+
+interface FetcherProps {
+  url?: string;
+  urls?: string[];
+  options?: RequestInit;
+  children: (state: any) => any;
+}
+
+const urlsEqual = (urls?: string[], nextURLs?: string[]) => {
+  if (!urls || !nextURLs) {
+    return false;
+  }
+  return JSON.stringify(urls) === JSON.stringify(nextURLs);
+};
+
+export class Fetcher extends Component<FetcherProps, FetcherState<any>> {
+  state = {
+    data: undefined,
+    error: undefined,
+  };
+
+  componentDidMount() {
+    const { url, urls, options } = this.props;
+    if (urls && urls.length) {
+      this.handleResponse(async () => await Promise.all(urls.map(this.get(options)).filter(Boolean)));
+    } else if (url) {
+      this.handleResponse(() => this.get(options)(url));
+    } else {
+      throw new Error('URL/s is Missing');
+    }
+  }
+
+  componentDidUpdate(nextProps: FetcherProps) {
+    const { url: nextURL, urls: nextURLS, options } = nextProps;
+    const { url, urls } = this.props;
+    if (nextURLS && !urlsEqual(urls, nextURLS)) {
+      this.handleResponse(async () => await Promise.all(nextURLS.map(this.get(options)).filter(Boolean)));
```

Comment:
If for some reason undefined or null appears, filter(Boolean) will remove them.
"Wouldn't one always expect to get back data for each URL" - yes, I guess so but I want to be sure that nothing unexpected will happen after that.

Reply:
Does this mean that if for some reason an api returns a falsey value then we just silently discard the result and therefore change the ordering of the results? If so that seems very bad.

Reply:
Yes, maybe it's good the consumer of the response to be aware of that. Will remove it from here and will check into component instead, thanks!